code/iaas/logic/src/main/java/io/cattle/platform/process/host/HostActivate.java

```diff
@@ -1,19 +1,65 @@
 package io.cattle.platform.process.host;
 
+import io.cattle.iaas.cluster.process.lock.ClusterLock;
+import io.cattle.iaas.cluster.service.ClusterManager;
+import io.cattle.platform.core.constants.ClusterConstants;
+import io.cattle.platform.core.constants.InstanceConstants;
 import io.cattle.platform.core.model.Agent;
 import io.cattle.platform.core.model.Host;
+import io.cattle.platform.core.model.Instance;
 import io.cattle.platform.engine.handler.HandlerResult;
 import io.cattle.platform.engine.process.ProcessInstance;
 import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.lock.LockCallback;
+import io.cattle.platform.lock.LockManager;
+import io.cattle.platform.object.resource.ResourceMonitor;
+import io.cattle.platform.object.resource.ResourcePredicate;
 import io.cattle.platform.process.base.AbstractDefaultProcessHandler;
 
+import javax.inject.Inject;
 import javax.inject.Named;
 
 @Named
 public class HostActivate extends AbstractDefaultProcessHandler {
 
+    @Inject
+    ResourceMonitor resourceMonitor;
+
+    @Inject
+    ClusterManager clusterManager;
+
+    @Inject
+    LockManager lockManager;
+
     @Override
     public HandlerResult handle(final ProcessState state, ProcessInstance process) {
+        if (objectManager.isKind(state.getResource(), ClusterConstants.KIND)) {
+            return handleCluster(state, process);
+        } else {
+            return handleHost(state, process);
+        }
+    }
+
+    private HandlerResult handleCluster(ProcessState state, ProcessInstance process) {
+        final Host cluster = (Host)state.getResource();
+        lockManager.lock(new ClusterLock(cluster), new LockCallback<Instance>() {
```

Comment:
Could probably use `LockCallbackNoReturn`, but not a blocker. Up to you if you want to change.